### ISSUE TITLE
Consider non-PHP webmails

### ISSUE THREAD

**Issue by nextgens** [Mailu/Mailu#2540]
They are lighter than running an HTTPd & FPM and potentially a lot easier to secure.

https://sr.ht/~migadu/alps/ was suggested by @jvoisin

**Comment by elandorr** [comment]
I'd be interested. I like rainloop, but resource savings are always appreciated. One benefit of the mailu stack is it being relatively lightweight. E.g. 'mailcow' is obscenely bloated, with the authors mocking everyone not rich enough to throw huge chunks of money at it.

PHP itself is fine really, but a single binary doing exactly one job sounds enticing. I don't know how much of that is just hype, though. If it's standalone, it has to recreate everything PHP-FPM offers, reinventing the wheel. A lot of these trendy golang tools are not very long lived or well thought out, while rainloop is proven.

I don't know the security implications. Are there even any substantial ones, given it's containerized? 

Another aspect to consider is that rainloop allows adding/editing sieve filters. The alps page doesn't list much, does it have that?
Rainloop also allows selecting identity, i.e. when using aliases, this is crucial.

If it doesn't take time from other issues and is enjoyable to mess with to someone, sounds like a good task!

**Comment by nextgens** [comment]
If you want to try it out, here is the Dockerfile I use:

```
ARG DISTRO=alpine:3.17.0

FROM $DISTRO as build

RUN set -euxo pipefail \
  ; apk add --no-cache go git \
  ; git clone --depth 1 https://git.sr.ht/~migadu/alps \
  ; cd alps/cmd/alps \
  ; go build

FROM $DISTRO
COPY --from=build /alps/ /alps/

EXPOSE 1323/tcp
HEALTHCHECK CMD curl -skfLo /dev/null http://localhost:1323/
WORKDIR /alps
USER nobody:nogroup

CMD /alps/cmd/alps/alps -theme alps imaps://test.mailu.io:993 smtps://test.mailu.io:465
```

The image is 25.7MB; RAM usage hovering around ~17MB.

All that's missing is the glue and a custom plugin to handle SSO.

**Comment by elandorr** [comment]
@nextgens Thank you, I appreciate it. Can't play with it right now, but thanks for sharing. Those numbers look interesting.

The current mailu/rainloop:1.9.43 shows between 0-15MB RAM only too, though. But the image is bigger: 118MB.

I wondered if it increases substantially during activity, but I can't login right now. It redirects a couple times, and goes back to the SSO, logging me out of the admin area as well. I forgot about the webmail when making a strict CSP to blacklist 3rd parties :O.

If anyone is searching CSP / content-security-policy in this repo: 
This is the strictest that lets mailu-admin+rainloop work:
```
default-src 'none'; connect-src 'self'; font-src 'self'; form-action 'self'; frame-src 'self'; img-src 'self' data:; script-src 'self' 'unsafe-eval' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; base-uri 'self';  manifest-src 'self'; frame-ancestors 'self';
```
This will block all 3rd parties even if the user selects 'load images', but I consider that a feature. If he still wants to, he can connect via IMAP.

Fixed.
So even when I reload wildly and click stuff, it doesn't go above 30MB RAM. Would be only about saving storage/potential security benefits then, which is nice too, of course

**Comment by acran** [comment]
Given that it is actually not that hard to integrate any arbitrary webmailer I think it would be a good idea to just document how to integrate an alternate webmailer. That way the users could choose and pick their preferred webmailer even if it's not officially supported by mailu (but instead by someone else in the community). mailu then only needs to support the interface and its documentation.

As far as I can tell the integration consists of two main components:

1. supply the webmailer with the correct configuration (host, ports etc.)
2. integrate login with SSO

Where 1. doesn't even need any special customization but only supplying the webmailer with a more or less static configuration with the right values. For 2. the latter needs to use the internal ports though - i.e. SMTP on `front:10025`, IMAP on `front:10143`.

Then 2. is done by using `front` as reverse proxy which supplies the users credentials as HTTP headers `X-Remote-User` and `X-Remote-User-Token`. Using them probably requires customizing the webmailer, e.g. with a custom auth plugin like it is done for roundcube. Also the webmailer needs to properly support the `$WEB_WEBMAIL` path prefix in the URL.

---

Currently though this integration is done in `front` by [this block in the `nginx.conf` template](https://github.com/Mailu/Mailu/blob/master/core/nginx/conf/nginx.conf#L195-L225) which evaluates to something like this (for `WEB_WEBMAIL=/webmail`):
~~~nginx
      location /webmail {
        rewrite ^(/webmail)$ $1/ permanent;
        rewrite ^/webmail/(.*) /$1 break;
        include /etc/nginx/proxy.conf;
        auth_request /internal/auth/user;
        error_page 403 @sso_login;
        proxy_pass http://$webmail;
      }

      location /webmail/sso.php {
        rewrite ^(/webmail)$ $1/ permanent;
        rewrite ^/webmail/(.*) /$1 break;
        include /etc/nginx/proxy.conf;
        auth_request /internal/auth/user;
        auth_request_set $user $upstream_http_x_user;
        auth_request_set $token $upstream_http_x_user_token;
        proxy_set_header X-Remote-User $user;
        proxy_set_header X-Remote-User-Token $token;
        error_page 403 @sso_login;
        proxy_pass http://$webmail;
      }
~~~

I.e. the auth token is only added as HTTP header for the `/sso.php` endpoint. Otherwise the `location` blocks are actually identical.

Is there a good reason for this distinction or couldn't we just merge the `location` blocks and always pass the token through to the webmailer?

Alternatively the `/sso.php` path for login should be configurable (e.g. `WEB_WEBMAIL_SSO_LOCATION` defaulting to `$WEB_WEBMAIL/sso.php`) to not needlessly complicate integration with non-PHP webmailers.

**Comment by nextgens** [comment]
> Is there a good reason for this distinction

Yes, security. This is the only script where we expose the temporary credentials.

**Comment by Willow0349** [comment]
I have made a proof of concept with alps behind mailu's nginx and with integration for login:
https://github.com/Willow0349/alps

There is a How-To in the README.md
